driver/src/price_finding/stablex_naive_solver.rs

```diff
@@ -0,0 +1,242 @@
+            .nth(0)
+            .unwrap_or_else(|| Solution::trivial(orders.len())))
+    }
+}
+
+const BASE_UNIT: u128 = 1_000_000_000_000_000_000u128;
+const BASE_PRICE: u128 = BASE_UNIT;
+const FEE_DENOM: u128 = 1000;
+
+/// Returns an iterator over all unordered pairs of indices in slice.
+fn ipairs<T>(slice: &[T]) -> impl Iterator<Item = (usize, usize)> {
+    let len = slice.len();
+    (0..len - 1)
+        .map(move |i| (i + 1..len).map(move |j| (i, j)))
+        .flatten()
+}
+
+/// Returns true if the owner of an order has enough balance to fully fill the
+/// order; false otherwise.
+fn has_enough_balance(order: &Order, accounts: &AccountState) -> bool {
+    order.sell_amount <= accounts.read_balance(order.sell_token, order.account_id)
+}
+
+/// Gets the limit price of the non-fee token for the given order considering
+/// fees.
+fn order_limit_price(order: &Order) -> u128 {
+    use util::u128_to_u256 as u;
+
+    let (fee, other, add_fees) = if order.buy_token == 0 {
+        (order.buy_amount, order.sell_amount, true)
+    } else {
+        (order.sell_amount, order.buy_amount, false)
+    };
+
+    // upcast to 256 to avoid overflows
+    let price = util::u256_to_u128((u(fee) * u(BASE_PRICE)) / u(other));
+    // account for fees in price
+    if add_fees {
+        (price * FEE_DENOM) / (FEE_DENOM - 1)
+    } else {
+        (price * (FEE_DENOM - 1)) / FEE_DENOM
+    }
+}
+
+/// Calculate the average of two values, note that this takes extra care not to
+/// overflow.
+fn average(a: u128, b: u128) -> u128 {
+    let (min, max) = if a < b { (a, b) } else { (b, a) };
+    min + ((max - min) / 2)
+}
+
+/// Calculate the executed buy amount from the executed sell amount and the buy
+/// and sell prices of the traded tokens. This function returns a `None` if no
+/// value can be found such that `executed_sell_amount(result, pb, ps) == xs`.
+/// This function acts as an inverse to `executed_sell_amount`.
+fn executed_buy_amount(xs: u128, pb: u128, ps: u128) -> Option<u128> {
+    use util::u128_to_u256 as u;
+
+    let v = util::u256_to_u128((((u(xs) * u(ps)) / u(FEE_DENOM)) * u(FEE_DENOM - 1)) / u(pb));
+
+    // we need to account for rounding errors here, since this function is
+    // essentially an inverse of `executed_sell_amount`; so find the maximum
+    // error that `v` can have and find a value that works
+    // TODO(nlordell): verify the maths
+
+    macro_rules! return_if_correct {
+        ($v:expr) => {
+            if xs == executed_sell_amount($v, pb, ps) {
+                return Some($v);
+            }
+        };
+    }
+
+    return_if_correct!(v);
+
+    let delta = (ps / pb) + 1;
+    for d in 1..=delta {
+        return_if_correct!(v + d);
+        return_if_correct!(v - d);
+    }
+
+    None
+}
+
+/// Calculate the executed sell amount from the executed buy amount and the buy
+/// and sell prices of the traded tokens.
+fn executed_sell_amount(xb: u128, pb: u128, ps: u128) -> u128 {
```

Comment:
can we spell the variables out, I find it really hard to deal with xb, pb and so on. It takes a few more keystrokes to write, but will make it easier to read I believe.

Reply:
Yes... but... my... lines... so... long...